bgpd/bgpd.c

```diff
@@ -4902,6 +4917,43 @@ int peer_ebgp_multihop_unset(struct peer *peer)
 	return 0;
 }
 
+/* Set Open Policy Role and check its correctness */
+int peer_role_set(struct peer *peer, uint8_t role, bool strict_mode)
+{
+	if (peer->local_role == role) {
+		if (CHECK_FLAG(peer->flags, PEER_FLAG_STRICT_MODE) &&
+		    !strict_mode)
+			/* TODO: Is session restart needed if it was down? */
+			UNSET_FLAG(peer->flags, PEER_FLAG_STRICT_MODE);
+		if (!CHECK_FLAG(peer->flags, PEER_FLAG_STRICT_MODE) &&
+		    strict_mode) {
+			SET_FLAG(peer->flags, PEER_FLAG_STRICT_MODE);
+			/* Restart session to throw Role Mismatch Notification
+			 */
+			if (peer->remote_role == ROLE_UNDEFINED)
+				bgp_session_reset(peer);
+		}
+	} else {
+		if (peer->sort == BGP_PEER_IBGP &&
```

Comment:
Shouldn't we ignore setting the roles for iBGP?

Reply:
As the RFC says, we should not apply the standard roles (provider...peer) to iBGP. Since there is no way to distinguish between eBGP and iBGP during the VTY command (at least not during the autocomplete step), this check has been added here.

Between ignoring the internal role and providing a configuration exception, I prefer the exception:
- it's clear that iBGP roles are not yet supported, so you don't have wrong assumptions about applying OTC rules in your internal network;
- when adding iBGP roles (with new possible drafts) this message disappeared and this is a signal that iBGP roles have been supported.

Reply:
I was about we might just return it early before continuing like:

```c
if (peer->sort != BGP_PEER_IBGP)
  return;
```

Reply:
Agree with that. The previous condition was part of the deprecated code when the internal role had not yet been removed from the draft and such roles enumeration made sense.

Fixed with the latest commit.